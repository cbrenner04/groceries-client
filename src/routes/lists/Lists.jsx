import React from 'react';
import Async from 'react-async';
import PropTypes from 'prop-types';

import { fetchLists } from './utils';
import ListsContainer from './containers/ListsContainer';
import Loading from '../../components/Loading';
import UnknownError from '../error_pages/UnknownError';

function Lists(props) {
  return (
    <Async promiseFn={fetchLists} history={props.history}>
      <Async.Pending>
        <Loading />
      </Async.Pending>
      <Async.Fulfilled>
        {(data) => (
          <ListsContainer
            history={props.history}
            userId={data.userId}
            pendingLists={data.pendingLists}
            completedLists={data.completedLists}
            nonCompletedLists={data.nonCompletedLists}
            currentUserPermissions={data.currentUserPermissions}
          />
        )}
      </Async.Fulfilled>
      <Async.Rejected>
        <UnknownError />
      </Async.Rejected>
    </Async>
  );
}

Lists.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
    replace: PropTypes.func.isRequired,
    location: PropTypes.shape({
      pathname: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};

export default Lists;
