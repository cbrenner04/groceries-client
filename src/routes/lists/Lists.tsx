import React, { type ReactNode } from 'react';
import Async, { type PromiseFn } from 'react-async';
import { useNavigate } from 'react-router';

import Loading from 'components/Loading';
import type { IList, TUserPermissions } from 'typings';

import { fetchLists } from './utils';
import ListsContainer from './containers/ListsContainer';
import UnknownError from '../error_pages/UnknownError';

interface IFulfilledLists {
  userId: string;
  pendingLists: IList[];
  completedLists: IList[];
  incompleteLists: IList[];
  currentUserPermissions: TUserPermissions;
}

const Lists: React.FC = (): React.JSX.Element => {
  const navigate = useNavigate();

  return (
    <Async promiseFn={fetchLists as unknown as PromiseFn<void>} navigate={navigate}>
      <Async.Pending>
        <Loading />
      </Async.Pending>
      <Async.Fulfilled>
        {(data: IFulfilledLists | undefined): ReactNode => {
          // Handle the case where data might be undefined (e.g., when fetchLists returns undefined due to an error)
          if (!data) {
            return <UnknownError />;
          }

          return (
            <ListsContainer
              userId={data.userId}
              pendingLists={data.pendingLists}
              completedLists={data.completedLists}
              incompleteLists={data.incompleteLists}
              currentUserPermissions={data.currentUserPermissions}
            />
          );
        }}
      </Async.Fulfilled>
      <Async.Rejected>
        <UnknownError />
      </Async.Rejected>
    </Async>
  );
};

export default Lists;
