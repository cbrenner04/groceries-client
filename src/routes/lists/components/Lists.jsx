import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

import List from './List';

const Lists = props => (
  <div>
    { props.pendingLists.length > 0 &&
      (
        <div>
          <p>These lists have been shared with you but you have not accepted the invitation.</p>
          <div className="list-group">
            {
              props.pendingLists.map(list => (
                <List
                  userId={props.userId}
                  list={list}
                  key={list.id}
                  onListAcceptance={props.onAccept}
                  onListRejection={props.onReject}
                />
              ))
            }
          </div>
          <hr />
        </div>
      )
    }
    <h1>Your Lists</h1>
    <p>
      These are lists you&apos;ve created or you&apos;ve accepted an
      invitation from someone else.
    </p>
    <div className="list-group">
      {
        props.nonCompletedLists.map(list => (
          <List
            userId={props.userId}
            list={list}
            key={list.id}
            onListDeletion={props.onListDelete}
            onListCompletion={props.onListCompletion}
            completed={list.completed}
            accepted
          />
        ))
      }
    </div>
    <br />
    <h2>Completed Lists</h2>
    <p>
      These are the completed lists most recently created.&nbsp;
      <Link to="/completed_lists">See all completed lists here.</Link>&nbsp;
      Previously refreshed lists are marked with an asterisk (*).
    </p>
    <div className="list-group">
      {
        props.completedLists.map(list => (
          <List
            userId={props.userId}
            list={list}
            key={list.id}
            onListDeletion={props.onListDelete}
            completed={list.completed}
            onListRefresh={props.onListRefresh}
            accepted
          />
        ))
      }
    </div>
  </div>
);

Lists.propTypes = {
  userId: PropTypes.number.isRequired,
  completedLists: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    created_at: PropTypes.string.isRequired,
    completed: PropTypes.bool.isRequired,
    users_list_id: PropTypes.number,
    owner_id: PropTypes.number,
  }).isRequired).isRequired,
  nonCompletedLists: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    created_at: PropTypes.string.isRequired,
    completed: PropTypes.bool.isRequired,
    users_list_id: PropTypes.number,
    owner_id: PropTypes.number,
  }).isRequired).isRequired,
  pendingLists: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    created_at: PropTypes.string.isRequired,
    completed: PropTypes.bool.isRequired,
    users_list_id: PropTypes.number,
    owner_id: PropTypes.number,
  }).isRequired).isRequired,
  onAccept: PropTypes.func.isRequired,
  onReject: PropTypes.func.isRequired,
  onListCompletion: PropTypes.func.isRequired,
  onListDelete: PropTypes.func.isRequired,
  onListRefresh: PropTypes.func.isRequired,
};

export default Lists;
