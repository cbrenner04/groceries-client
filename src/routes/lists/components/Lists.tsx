import React, { type ReactElement, type ReactNode } from 'react';
import { Button } from 'react-bootstrap';

import type { IList } from 'typings';

export interface IListsProps {
  title: ReactElement;
  multiSelect: boolean;
  selectedLists: IList[];
  setSelectedLists: (lists: IList[]) => void;
  setMultiSelect: (select: boolean) => void;
  children: ReactNode[];
}

const Lists: React.FC<IListsProps> = (props): React.JSX.Element => {
  return (
    <div className="mb-4">
      <div className="clearfix">
        <div className="mx-auto float-start">{props.title}</div>
        <Button
          variant="link"
          className="mx-auto float-end"
          onClick={(): void => {
            if (props.multiSelect && props.selectedLists.length > 0) {
              props.setSelectedLists([]);
            }
            props.setMultiSelect(!props.multiSelect);
          }}
        >
          {props.multiSelect ? 'Hide' : ''} Select
        </Button>
      </div>
      {props.children}
    </div>
  );
};

export default Lists;
