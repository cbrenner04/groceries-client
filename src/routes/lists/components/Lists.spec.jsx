import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Lists from './Lists';

function setup(suppliedProps = {}) {
  const user = userEvent.setup();
  const defaultProps = {
    title: <div />,
    children: [<div key="1" />],
    multiSelect: false,
    selectedLists: [],
    setSelectedLists: jest.fn(),
    setMultiSelect: jest.fn(),
  };
  const props = { ...defaultProps, ...suppliedProps };
  const component = render(<Lists {...props} />);

  return { ...component, props, user };
}

describe('Lists', () => {
  it('sets multiSelect to true when select is clicked', async () => {
    const { container, findAllByText, props, user } = setup({ multiSelect: false, selectedLists: [] });

    await user.click((await findAllByText('Select'))[0]);

    expect(container).toMatchSnapshot();
    expect(props.setMultiSelect).toHaveBeenCalledWith(true);
  });

  it('sets multiSelect to false and clears selectedLists when Hide Select is clicked', async () => {
    const { container, findAllByText, props, user } = setup({
      multiSelect: true,
      selectedLists: [
        {
          id: 'id1',
          name: 'foo',
          type: 'GroceryList',
          created_at: new Date('05/27/2020').toISOString(),
          completed: true,
          users_list_id: 'id1',
          owner_id: 'id1',
          refreshed: false,
        },
      ],
    });

    await user.click((await findAllByText('Hide Select'))[0]);

    expect(container).toMatchSnapshot();
    expect(props.setMultiSelect).toHaveBeenCalledWith(false);
    expect(props.setSelectedLists).toHaveBeenCalledWith([]);
  });
});
