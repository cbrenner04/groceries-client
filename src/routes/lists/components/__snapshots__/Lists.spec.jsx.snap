// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lists does not render pending lists when they do not exist 1`] = `
<div>
  <h1>
    Your Lists
  </h1>
  <p>
    These are lists you've created or you've accepted an invitation from someone else.
  </p>
  <div
    class="clearfix"
  >
    <button
      class="mx-auto float-right btn btn-link"
      type="button"
    >
      
       Select
    </button>
  </div>
  <div
    class="list-group"
  >
    <div
      class="accepted-list list-group-item"
      data-test-class="non-completed-list"
      data-test-id="list-1"
      style="display: block;"
    >
      <div
        class="row"
      >
        <div
          class="col-sm-12 col-12"
        >
          <div
            class="row"
          >
            <div
              class="pt-1 col-md-6"
            >
              <a
                class="router-link"
                href="/lists/1"
              >
                <h5
                  class="mb-1"
                >
                  <i
                    class="fa fa-shopping-bag text-info mr-3"
                  />
                  foo
                </h5>
              </a>
            </div>
            <div
              class="pt-1 col-md-4"
            >
              <small
                class="text-muted"
              >
                May 27 2020, 12:00:00 am
              </small>
            </div>
            <div
              class="col-md-2"
            >
              <div
                class="float-right btn-group"
                role="group"
              >
                <button
                  class="p-0 mr-3 btn btn-link"
                  data-test-id="incomplete-list-complete"
                  style="opacity: 1;"
                  type="button"
                >
                  <i
                    class="fa fa-check fa-2x text-success"
                  />
                </button>
                <button
                  class="p-0 mr-3 btn btn-link"
                  data-test-id="incomplete-list-share"
                  style="pointer-events: auto; opacity: 1;"
                  type="button"
                >
                  <i
                    class="fa fa-users fa-2x text-primary"
                  />
                </button>
                <a
                  class="p-0 mr-3"
                  data-test-id="incomplete-list-edit"
                  href="/lists/1/edit"
                  style="pointer-events: auto; opacity: 1;"
                >
                  <i
                    class="fa fa-edit fa-2x text-warning"
                  />
                </a>
                <button
                  class="p-0 btn btn-link"
                  data-test-id="incomplete-list-trash"
                  type="button"
                >
                  <i
                    class="fa fa-trash fa-2x text-danger"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <br />
  <h2>
    Completed Lists
  </h2>
  <p>
    These are the completed lists most recently created. 
    <a
      href="/completed_lists"
    >
      See all completed lists here.
    </a>
      Previously refreshed lists are marked with an asterisk (*).
  </p>
  <div
    class="list-group"
  />
</div>
`;

exports[`Lists renders complete lists 1`] = `
<div>
  <h1>
    Your Lists
  </h1>
  <p>
    These are lists you've created or you've accepted an invitation from someone else.
  </p>
  <div
    class="clearfix"
  >
    <button
      class="mx-auto float-right btn btn-link"
      type="button"
    >
      
       Select
    </button>
  </div>
  <div
    class="list-group"
  />
  <br />
  <h2>
    Completed Lists
  </h2>
  <p>
    These are the completed lists most recently created. 
    <a
      href="/completed_lists"
    >
      See all completed lists here.
    </a>
      Previously refreshed lists are marked with an asterisk (*).
  </p>
  <div
    class="list-group"
  >
    <div
      class="accepted-list list-group-item"
      data-test-class="completed-list"
      data-test-id="list-1"
      style="display: block;"
    >
      <div
        class="row"
      >
        <div
          class="col-sm-12 col-12"
        >
          <div
            class="row"
          >
            <div
              class="pt-1 col-md-6"
            >
              <a
                class="router-link"
                href="/lists/1"
              >
                <h5
                  class="mb-1"
                >
                  <i
                    class="fa fa-shopping-bag text-info mr-3"
                  />
                  foo
                </h5>
              </a>
            </div>
            <div
              class="pt-1 col-md-4"
            >
              <small
                class="text-muted"
              >
                May 27 2020, 12:00:00 am
              </small>
            </div>
            <div
              class="col-md-2"
            >
              <div
                class="float-right btn-group"
                role="group"
              >
                <button
                  class="p-0 mr-3 btn btn-link"
                  data-test-id="complete-list-refresh"
                  style="opacity: 1;"
                  type="button"
                >
                  <i
                    class="fa fa-redo fa-2x text-primary"
                  />
                </button>
                <button
                  class="p-0 btn btn-link"
                  data-test-id="complete-list-trash"
                  type="button"
                >
                  <i
                    class="fa fa-trash fa-2x text-danger"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Lists renders incomplete lists 1`] = `
<div>
  <h1>
    Your Lists
  </h1>
  <p>
    These are lists you've created or you've accepted an invitation from someone else.
  </p>
  <div
    class="clearfix"
  >
    <button
      class="mx-auto float-right btn btn-link"
      type="button"
    >
      
       Select
    </button>
  </div>
  <div
    class="list-group"
  >
    <div
      class="accepted-list list-group-item"
      data-test-class="non-completed-list"
      data-test-id="list-1"
      style="display: block;"
    >
      <div
        class="row"
      >
        <div
          class="col-sm-12 col-12"
        >
          <div
            class="row"
          >
            <div
              class="pt-1 col-md-6"
            >
              <a
                class="router-link"
                href="/lists/1"
              >
                <h5
                  class="mb-1"
                >
                  <i
                    class="fa fa-shopping-bag text-info mr-3"
                  />
                  foo
                </h5>
              </a>
            </div>
            <div
              class="pt-1 col-md-4"
            >
              <small
                class="text-muted"
              >
                May 27 2020, 12:00:00 am
              </small>
            </div>
            <div
              class="col-md-2"
            >
              <div
                class="float-right btn-group"
                role="group"
              >
                <button
                  class="p-0 mr-3 btn btn-link"
                  data-test-id="incomplete-list-complete"
                  style="opacity: 1;"
                  type="button"
                >
                  <i
                    class="fa fa-check fa-2x text-success"
                  />
                </button>
                <button
                  class="p-0 mr-3 btn btn-link"
                  data-test-id="incomplete-list-share"
                  style="pointer-events: auto; opacity: 1;"
                  type="button"
                >
                  <i
                    class="fa fa-users fa-2x text-primary"
                  />
                </button>
                <a
                  class="p-0 mr-3"
                  data-test-id="incomplete-list-edit"
                  href="/lists/1/edit"
                  style="pointer-events: auto; opacity: 1;"
                >
                  <i
                    class="fa fa-edit fa-2x text-warning"
                  />
                </a>
                <button
                  class="p-0 btn btn-link"
                  data-test-id="incomplete-list-trash"
                  type="button"
                >
                  <i
                    class="fa fa-trash fa-2x text-danger"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <br />
  <h2>
    Completed Lists
  </h2>
  <p>
    These are the completed lists most recently created. 
    <a
      href="/completed_lists"
    >
      See all completed lists here.
    </a>
      Previously refreshed lists are marked with an asterisk (*).
  </p>
  <div
    class="list-group"
  />
</div>
`;

exports[`Lists renders pending lists when they exist 1`] = `
<div>
  <p>
    These lists have been shared with you but you have not accepted the invitation.
  </p>
  <div
    class="list-group"
  >
    <div
      class="pending-list list-group-item"
      data-test-class="pending-list"
      data-test-id="list-1"
      style="display: block;"
    >
      <div
        class="row"
      >
        <div
          class="col-sm-12 col-12"
        >
          <div
            class="row"
          >
            <div
              class="pt-1 col-md-6"
            >
              <h5
                class="mb-1"
              >
                <i
                  class="fa fa-shopping-bag text-info mr-3"
                />
                foo
              </h5>
            </div>
            <div
              class="pt-1 col-md-4"
            >
              <small
                class="text-muted"
              >
                May 27 2020, 12:00:00 am
              </small>
            </div>
            <div
              class="col-md-2"
            >
              <div
                class="float-right btn-group"
                role="group"
              >
                <button
                  class="p-0 mr-3 btn btn-link"
                  data-test-id="pending-list-accept"
                  type="button"
                >
                  <i
                    class="fa fa-check fa-2x text-success"
                  />
                </button>
                <button
                  class="p-0 btn btn-link"
                  data-test-id="pending-list-trash"
                  type="button"
                >
                  <i
                    class="fa fa-trash fa-2x text-danger"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr />
  <h1>
    Your Lists
  </h1>
  <p>
    These are lists you've created or you've accepted an invitation from someone else.
  </p>
  <div
    class="clearfix"
  >
    <button
      class="mx-auto float-right btn btn-link"
      type="button"
    >
      
       Select
    </button>
  </div>
  <div
    class="list-group"
  />
  <br />
  <h2>
    Completed Lists
  </h2>
  <p>
    These are the completed lists most recently created. 
    <a
      href="/completed_lists"
    >
      See all completed lists here.
    </a>
      Previously refreshed lists are marked with an asterisk (*).
  </p>
  <div
    class="list-group"
  />
</div>
`;

exports[`Lists sets multiSelect to false and clears selectedLists when Hide Select is clicked 1`] = `
<div>
  <h1>
    Your Lists
  </h1>
  <p>
    These are lists you've created or you've accepted an invitation from someone else.
  </p>
  <div
    class="clearfix"
  >
    <button
      class="mx-auto float-right btn btn-link"
      type="button"
    >
      Hide
       Select
    </button>
  </div>
  <div
    class="list-group"
  />
  <br />
  <h2>
    Completed Lists
  </h2>
  <p>
    These are the completed lists most recently created. 
    <a
      href="/completed_lists"
    >
      See all completed lists here.
    </a>
      Previously refreshed lists are marked with an asterisk (*).
  </p>
  <div
    class="list-group"
  />
</div>
`;

exports[`Lists sets multiSelect to true when select is clicked 1`] = `
<div>
  <h1>
    Your Lists
  </h1>
  <p>
    These are lists you've created or you've accepted an invitation from someone else.
  </p>
  <div
    class="clearfix"
  >
    <button
      class="mx-auto float-right btn btn-link"
      type="button"
    >
      
       Select
    </button>
  </div>
  <div
    class="list-group"
  />
  <br />
  <h2>
    Completed Lists
  </h2>
  <p>
    These are the completed lists most recently created. 
    <a
      href="/completed_lists"
    >
      See all completed lists here.
    </a>
      Previously refreshed lists are marked with an asterisk (*).
  </p>
  <div
    class="list-group"
  />
</div>
`;
